### ISSUE TITLE
Add "group" to lgb.cv for non-rank objectives

### ISSUE THREAD

**Issue by JoshuaC3** [Microsoft/LightGBM#1541]
Lgb.cv accepts the group kwarg but only seems to use it in lambdarank objectives. See here: 
https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py

For more details on sklearns GroupKFold (use in lambdarank) see here: 
http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupKFold.html

_p.s._ will tidy u

**Comment by guolinke** [comment]
welcome to open the PR. We will help you after that : )

**Comment by JoshuaC3** [comment]
@guolinke - reading over the code it seems that there might be a nicer way to include GroupKFold, and some of the other [sklearn model selection cross validation iterators](http://scikit-learn.org/stable/modules/cross_validation.html#cross-validation-iterators).

The LightGBM Python package uses the [sklearn CV classes](https://github.com/Microsoft/LightGBM/blob/5fe2bdd7eae1fd2a10b2411c2f03fa477064a174/python-package/lightgbm/compat.py#L105) for Stratified and GroupKFold. Rather than adding code to allow _just_ GroupKFold, it would be a similar amount of work to pass an _uninitialised_ `sk.model_selection` object to `lgb.cv` and use this. This would also give us: GroupKFold, Time Series Split, LeaveOneGroupOut.

Even better, for a similar amount of work, we could pass an initialised `sk.model_selection` object. This would allow users to use _any_ of the [sklearn model selection cross validation iterators](http://scikit-learn.org/stable/modules/cross_validation.html#cross-validation-iterators). This would be my preferred option. The python code for this might look something like,

```
def _make_n_folds(full_data, folds, nfold, params, seed, fpreproc=None, stratified=True, shuffle=True):
    """
    Make an n-fold list of Booster from random indices.
    """
    full_data = full_data.construct()
    num_data = full_data.num_data()
    if folds is not None:
        if not hasattr(folds, '__iter__'):
            raise AttributeError("folds should be a generator or iterator of (train_idx, test_idx)")
        elif  hasattr(folds, 'split'):
            if groups is not None:
                group_info = full_data.get_group().astype(int)
                flatted_group = np.repeat(range_(len(group_info)), repeats=group_info)
            folds = skf.split(X=np.zeros(num_data), y=full_data.get_label(), groups=flatted_group)
```

This said, we can currently get this functionality using sklearn outside of `lgb.cv` using,

```
folds = sk.model_selection.GroupKFold().split(X, y, groups)
# other operations on X, y, data, etc
lgb.cv(params, data, folds=folds)
```

the new code would look like,

```
folds_object = sk.model_selection.GroupKFold()
# other operations on X, y, data, etc
lgb.cv(params, data, folds=folds_object )
```

The main advantage of this would be that `folds` would be using `X` and `y` before they are passed to an `lgb.Dataset` and so could be a different shape, order, group etc if X and y are operated on before training. With the _new_ version, `folds_object` would be guaranteed to split on the `lgb.Dataset`.

In summary, we gain only a small amount a functionality by implementing only GroupKFold. Should we instead implement the initialised, safer sklearn split option for `folds`, despite the functionality currently being achievable through another similar means?

**Comment by StrikerRUS** [comment]
+1 for supporting sklearn CV iterator. Please feel free to create a PR!

Just one minor remark after a brief glance at the future code prototype:
```
elif hasattr(folds, 'split'):
     if not SKLEARN_INSTALLED:
        raise LightGBMError('Some clever error about scikit-learn installation requirement')
```
